rtrlib/aspa/aspa_private.h

```diff
@@ -32,8 +46,89 @@
  * @param notify_dst A boolean value determining whether to notify the destination tables' clients.
  * @param notify_src A boolean value determining whether to notify the source tables' clients.
  */
-int aspa_table_src_move(struct aspa_table *dst, struct aspa_table *src, struct rtr_socket *rtr_socket, bool notify_dst,
-			bool notify_src);
+enum aspa_status aspa_table_src_replace(struct aspa_table *dst, struct aspa_table *src, struct rtr_socket *rtr_socket,
+					bool notify_dst, bool notify_src);
+
+// MARK: - Updating
+
+/**
+ * @brief An enum describing the type of operation the ASPA table should perform using any given ASPA record.
+ */
+enum aspa_operation_type {
+	/** The existing record, identified by its customer ASN, shall be withdrawn from the ASPA table. */
+	ASPA_REMOVE = 0,
+
+	/** The new record, identified by its customer ASN, shall be added to the ASPA table. */
+	ASPA_ADD = 1
+};
+
+/**
+ * @brief A struct describing a specific type of operation that should be performed using the attached ASPA record.
+ * @param index A value uniquely identifying this operation's position within the array of operations.
+ * @param type The operation's type.
+ * @param record The record that should be added or removed.
+ */
+struct aspa_update_operation {
+	size_t index;
+	enum aspa_operation_type type;
+	struct aspa_record record;
+};
+
+/**
+ * @brief ASPA update finalization arguments
+ * @param unused_provider_arrays Array of pointers to now unused provider sets.
+ * @param unused_provider_array_len Number of unused provider sets.
+ */
+struct aspa_update_finalization_args {
+	uint32_t **unused_provider_arrays;
+	size_t unused_provider_array_len;
+#ifndef ASPA_UPDATE_IN_PLACE
```

Comment:
ohhhh god - this is the direct pathway to circle 9 of hell.

Reply:
do you mean using a macro here or the args struct itself? If it's really the macro bothering you, we will get rid of that once we figured out which option is faster during the interop testing. (Depending on these results, this struct is either going to be renamed to `aspa_update_cleanup_args` or remain like that.